### PR TITLE
[Model] Add Zyphra ZONOS2 (MoE TTS) support on shared TTS surfaces

### PR DESCRIPTION
**Pull request by luojiaxuan**
Supersedes #779, carrying GaokaiZhang's ZONOS2 integration forward onto the current shared TTS architecture. All original commits are preserved in the branch history; this PR adds the remaining shared-surface adoption on top and brings the branch up to date with `main`.

## What this adds over #779

#779 had already adopted most shared surfaces (M2 `TtsEngineBuilder`, M4a `ReferenceEncodeService`, M6 `StreamingVocoderBase`, M8 `CAPABILITIES`, `DeclarativeStateBase` pipeline state, shared checkpoint resolution). This PR closes the gaps that landed on `main` after that branch's last sync:

1. **Speaker encoder on the #1096 keyed tensor hook.** `SpeakerEncoder` hand-rolled `cache_key`, `store_artifact`, and `load_artifact`; it now subclasses `TensorReferenceEncodeHook` with declarative identity attributes and an `input_key` accessor. Storage/load semantics are unchanged (CPU float32 copy in, isolated copy out).
2. **Shared `store_state` in the stage factories.** The local `_store` helper duplicated `scheduling.pipeline_state.store_state`; the preprocessing and speaker-encode stages now use the shared helper.
3. **`Zonos2State` joins the shared round-trip contract test.** #779 registered the state only in the usage-contract check; the field-complete round-trip case list in `tests/unit_test/scheduling/test_pipeline_state.py` now covers it too.
4. **Merged current `main`** (post-#1103), resolving cleanly with no conflicts.

Everything model-local by design stays local per the #985 conventions: the MoE backbone, DAC vocoder mechanics, sampler, radix hashing, decode-state pool (T4 is still open upstream), and the vendored NeMo text normalization.

## Behavior class

Output-equivalent adaptation on top of #779's model addition. No math or scheduling semantics change in this PR's own commits.

## Tests

- `tests/unit_test/zonos2/` (pipeline, speaker cache, streaming vocoder, text-normalize cache), `tests/unit_test/scheduling/`, and `tests/unit_test/models/test_model_capabilities.py` in an sglang-omni container on an A6000 host: **156 passed**.
- `black --check` clean on all files this PR touches beyond #779.

## Related

- Supersedes #779 (please keep review credit with its author)
- Shared-surface conventions: #985

Co-authored-by: GaokaiZhang <gaokaiz@andrew.cmu.edu>

### CODE REVIEW

File: sglang_omni/models/zonos2/sglang_model.py [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+        self.speaker_projection = nn.Linear(cfg.speaker_lda_dim, cfg.dim, bias=True)
+        self.layers = nn.ModuleList(
+            [Zonos2DecoderLayer(cfg, i, quant_config) for i in range(cfg.n_layers)]
+        )
+        self.out_norm = nn.Parameter(torch.empty(cfg.dim))
+        self.multi_output = nn.Parameter(
+            torch.empty(self.audio_vocab * self.n_codebooks, cfg.dim)
+        )
+
+        max_bs = 1
```
[the remark this answers is not in the document]
Comment (luojiaxuan) [reply sglang_omni/models/zonos2/sglang_model.py]: Done, removed the overwritten max_bs initialization.

File: sglang_omni/models/zonos2/components/audio_codec.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            .to(device)
+        )
+        # note (Yue Yin): the DAC convs use legacy weight_norm (78 modules) that
+        # recomputes g*v/||v|| every forward. For inference the reparam is fixed,
+        # so fold it into the conv weight once at load -> ~2e-4 fp-reassoc on the
+        # output (WER-neutral, verified), fewer ops per (streaming) vocoder call.
+        for _m in _dac_model.modules():
+            try:
+                torch.nn.utils.remove_weight_norm(_m)
+            except (ValueError, RuntimeError):
```
[the remark this answers is not in the document]
Comment (luojiaxuan) [reply sglang_omni/models/zonos2/components/audio_codec.py:39]: Done, documented the expected weight-norm removal failures.

File: sglang_omni/models/zonos2/engine_builder.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+            sdir = os.path.join(src_root, vdir)
+            if not os.path.isdir(sdir):
+                continue
+            ddir = os.path.join(dst_root, vdir)
+            os.makedirs(ddir, exist_ok=True)
+            for fn in os.listdir(sdir):
+                dst = os.path.join(ddir, fn)
+                if not os.path.exists(dst):
+                    shutil.copy2(os.path.join(sdir, fn), dst)
+    except Exception:
```
[the remark this answers is not in the document]
Comment (luojiaxuan) [reply sglang_omni/models/zonos2/engine_builder.py:98]: Done, tuned MoE config installation failures now log a warning with traceback.

File: sglang_omni/models/zonos2/sglang_model.py [hunk 4]
```diff
@@ -0,0 +1,612 @@
...
+            else sd
+        )
+        fixed: dict[str, torch.Tensor] = {}
+        for k, v in sd.items():
+            if ".router.ent_denom" in k or ".router.normalized_entropy" in k:
+                continue
+            if ".parametrizations." in k and ".original" in k:
+                k = k.replace(".parametrizations.", ".").replace(".original", "")
+            fixed[k] = v
+        params = dict(self.named_parameters())
```
[the remark this answers is not in the document]
Comment (luojiaxuan) [reply sglang_omni/models/zonos2/sglang_model.py]: Done, removed the unused params mapping.

File: sglang_omni/models/zonos2/components/audio_codec.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+                pass
+        _dac_device = device
```
[the remark this answers is not in the document]
Comment (luojiaxuan) [reply sglang_omni/models/zonos2/components/audio_codec.py]: Done, folded the DAC model and device into one cache tuple and covered device changes.

File: sglang_omni/models/zonos2/components/streaming_vocoder.py [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+)
+from sglang_omni.proto import StagePayload
+from sglang_omni.scheduling.pipeline_state import build_usage
+from sglang_omni.scheduling.streaming_vocoder import (
+    StreamingVocoderBase,
+    resolve_initial_codec_chunk_frames,
+)
+from sglang_omni.utils.audio_payload import audio_waveform_payload
+
+logger = logging.getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (luojiaxuan) [reply sglang_omni/models/zonos2/components/streaming_vocoder.py]: Done, removed the unused streaming vocoder logger.

File: sglang_omni/models/zonos2/components/streaming_vocoder.py [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+# pulls hold this region back; the eos_frame-capped flush at stream_done emits
+# the true tail, keeping streamed length == one-shot length.
+_STREAM_EOS_GUARD_FRAMES = N_CODEBOOKS + 1
+
+
+def _get_vocoder(device: str) -> Zonos2DACVocoder:
+    global _vocoder, _vocoder_device
+    if _vocoder is None or _vocoder_device != device:
+        _vocoder = Zonos2DACVocoder(device=device)
+        _vocoder_device = device
```
[the remark this answers is not in the document]
Comment (luojiaxuan) [reply sglang_omni/models/zonos2/components/streaming_vocoder.py]: Done, folded the vocoder and device into one cache tuple and covered device changes.

File: sglang_omni/models/zonos2/stages.py [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+        cache_max_items=speaker_cache_max_items,
+        compile_forward=spk_compile,
+    )
+
+    def _speaker(payload: StagePayload) -> StagePayload:
+        state = Zonos2State.from_dict(payload.data)
+        if state.ref_audio is not None:
+            ref = ref_audio_to_encoder_input(state.ref_audio)
+            state.speaker_emb = encoder.encode(ref)
+            state.speaker_fingerprint = encoder.fingerprint()
```
Comment (SandyLuXY) [inline sglang_omni/models/zonos2/stages.py]: `fingerprint()` returns the shared encoder instance's most recent key, not the key for this `encode()` call. With `max_concurrency=4`, request A can block in the encoder, request B can overwrite `_last_fingerprint`, and A then stores B's key as its `Req.extra_key`, allowing speaker-conditioned KV state to be cached under the wrong voice. Please return the embedding and fingerprint from the same per-call normalized item (without mutable “last request” state) and add a controlled two-request interleaving test.

File: sglang_omni/models/zonos2/request_builders.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+    data = Zonos2SGLangRequestData(
+        input_ids=torch.tensor(row_keys, dtype=torch.long),
+        max_new_tokens=max_new,
+        output_ids=req.output_ids,
+        req=req,
+        prompt_rows=rows,
+        speaker_emb=speaker_emb,
+        speaker_position=speaker_position,
+        params=params,
+        generator=gen,
```
Comment (SandyLuXY) [inline sglang_omni/models/zonos2/request_builders.py]: This generator is dead state: nothing reads `data.generator`, both eager and frame-graph sampling call `torch.multinomial` through the global device RNG, and the public `seed` is therefore silently ignored. Please thread a per-request deterministic seed plus generation position/codebook through both sampling paths and test invariance to batch neighbors, or reject `seed` for ZONOS2 until that contract is supported.

File: sglang_omni/models/zonos2/config.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    stages: list[StageConfig] = Field(
+        default_factory=lambda: _stages(codec_device="cuda:0", speaker_device="cuda:0")
+    )
+
+
+class Zonos2MultiGPUPipelineConfig(Zonos2PipelineConfig):
+    """Offload codec + speaker encoder to cuda:1, leaving the AR engine alone on cuda:0."""
+
+    stages: list[StageConfig] = Field(
+        default_factory=lambda: _stages(codec_device="cuda:1", speaker_device="cuda:1")
```
Comment (SandyLuXY) [inline sglang_omni/models/zonos2/config.py]: This does not create a typed two-GPU deployment: `_stages()` still declares every stage as `gpu=0` in the single `"pipeline"` process, so the resolved placement contains only GPU 0 even though two factories later touch `cuda:1`. Please split the GPU-1 stages into an explicit process group, declare `gpu=1`, derive the device from the injected placement.

File: tests/test_model/test_zonos2_tts_ci.py [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+    )
+
+
+@pytest.fixture(scope="module")
+def wer_audio_dir(speed_artifacts: _SpeedArtifacts) -> str:
+    """Reuse speed-benchmark audio for WER after freeing the TTS server GPU."""
+    # zonos2_server (depended on by speed_artifacts) is module-scoped; it tears
+    # down and frees the GPU after the last test that needs it. Force a cleanup
+    # here so the Qwen3-ASR WER router has the GPU before it starts.
+    wait_for_gpu_memory_release()
```
Comment (SandyLuXY) [inline tests/test_model/test_zonos2_tts_ci.py]: This wait cannot free the TTS server: `zonos2_server` is a module-scoped transitive dependency of `speed_artifacts`, and `pytest --setup-plan` tears it down only after `qwen3_asr_wer_router` and the WER test finish. Please run generation inside a server context that exits before the artifact fixture returns (or otherwise split the scopes), then wait for release before starting ASR.

File: sglang_omni/models/zonos2/components/text_frontend.py [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+    seed: int | None = None
+    ignore_eos: bool = False
+    n_codebooks: int = N_CODEBOOKS
+    eoa_id: int = EOA_ID
+    repetition_window: int = 50
+    repetition_codebooks: int = 8
+
+
+# Server language codes -> NeMo text_normalization language packages.
+_SERVER_TO_NEMO_LANG: dict[str, str] = {
```
Comment (SandyLuXY) [inline sglang_omni/models/zonos2/components/text_frontend.py:94]: These keys cannot be produced by `/v1/audio/speech`: the shared validator canonicalizes values to names such as `"English"` and `"Chinese"`, so `normalize_text()` rejects every documented language before reaching NeMo (for example, validated `"Chinese"` is not a key here). Please map the canonical public names to the supported NeMo codes, define `Auto`/unsupported behavior, and add speech-request-to-prompt normalization tests for English and Chinese.

File: docs/cookbook/zonos2.md [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+URL, or a base64 **data URI** (`data:audio/wav;base64,<...>`, transcoded via `ffmpeg`):
+
+```json
+{"ref_audio": "data:audio/wav;base64,UklGR.....", "ref_text": "Transcript of the clip."}
+```
+
+### Streaming
+
+Set `"stream": true` to receive Server-Sent Events (SSE). Audio events carry base64-encoded WAV
+bytes in `audio.data`; the final metadata event has `audio: null`, followed by `data: [DONE]`.
```
Comment (SandyLuXY) [inline docs/cookbook/zonos2.md]: This is not the speech endpoint's streaming contract: the example is rejected because `stream=true` requires `response_format="pcm"`, and a valid request returns raw PCM bytes rather than SSE/base64 WAV events or `[DONE]`. Please add the PCM format to the request, document the 44.1 kHz PCM stream and response headers, update the parameter table, and show saving/converting the raw bytes.

File: docs/cookbook/zonos2.md [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+  --port 8000
+```
+
+## Synthesizing Speech
+
+### Voice Cloning
+
+ZONOS2 clones a voice from a reference clip. The `references` field accepts `audio_path`
+(a local path, HTTP URL, or base64 data URI) and `text` (the transcript of that clip). Supplying
+the transcript materially improves cloning quality.
```
Comment (SandyLuXY) [inline docs/cookbook/zonos2.md]: ZONOS2 never consumes this transcript: `ref_text` is resolved into `Zonos2State` but has no reader anywhere in the model pipeline, so it cannot materially improve cloning quality. Please document it as an accepted-but-unused shared API field (and remove the same claim under Known Limitations), or implement and test an actual transcript-conditioned path.

File: tests/test_model/test_zonos2_tts_ci.py [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+#
+# Measured reference (NOT the gate) from a single dev H100 80GB run on this
+# branch (Track B rebased onto ZONOS2 = Track A + FP8-off + loop-guard), c=16,
+# seed-tts-eval-50 EN:
+#   rtf_mean=0.8212  throughput_qps=5.167  latency_mean_s=2.827
+# output_tok_per_req_s is unavailable via --use-existing-server (no per-request
+# engine_time_s); the in-harness managed-server path the test uses populates it,
+# so calibrate there. WER sanity check (whisper-large-v3, not the Qwen3-ASR
+# gate judge): corpus 0.89%, no >50% outliers.
+_VC_NON_STREAM_P95: dict[int, dict[str, float]] | None = None
```
Comment (SandyLuXY) [inline tests/test_model/test_zonos2_tts_ci.py]: Leaving this as `None` makes `test_voice_cloning_non_streaming` skip `assert_speed_thresholds()`, so any RTF/throughput/latency regression passes while the test is presented as a speed gate. Please populate a repeat-derived baseline.

**Review by SandyLuXY — CHANGES_REQUESTED** [review]

File: sglang_omni/models/zonos2/config.py (hunk 10 above)
Comment (luojiaxuan) [reply sglang_omni/models/zonos2/config.py]: Done, the multi-GPU variant now places speaker encoding and vocoding in an explicit GPU-1 process and derives devices from placement.

File: docs/cookbook/zonos2.md (hunk 13 above)
Comment (luojiaxuan) [reply docs/cookbook/zonos2.md]: Done, the streaming example now requests raw 44.1 kHz PCM and shows saving and converting it.

File: sglang_omni/models/zonos2/request_builders.py (hunk 9 above)
Comment (luojiaxuan) [reply sglang_omni/models/zonos2/request_builders.py]: Done, ZONOS2 now rejects seed until per-request sampling is supported, and the dead generator state is removed.

File: tests/test_model/test_zonos2_tts_ci.py (hunk 15 above)
Comment (luojiaxuan) [reply tests/test_model/test_zonos2_tts_ci.py]: Done, the speed gate now uses the recorded cold H100 calibration envelope instead of skipping thresholds.

File: sglang_omni/models/zonos2/stages.py (hunk 8 above)
Comment (luojiaxuan) [reply sglang_omni/models/zonos2/stages.py]: Done, the encoder now returns the embedding and fingerprint from one normalized input, with a controlled interleaving test.

File: docs/cookbook/zonos2.md (hunk 14 above)
Comment (luojiaxuan) [reply docs/cookbook/zonos2.md]: Done, the docs now state that ref_text is accepted but unused by ZONOS2.

File: tests/test_model/test_zonos2_tts_ci.py (hunk 15 above)
Comment (luojiaxuan) [reply tests/test_model/test_zonos2_tts_ci.py]: Updated with three cold H100 runs. The gate now uses the worst observed QPS, latency, and RTF.

**Comment by SandyLuXY** [comment]
Benchmarking results on 1xH100: 

## Setup

- Candidate: `@23ab18d`
- Model/checkpoint: `Zyphra/zonos2`
- Dataset/eval suite: `zhaochenyang20/seed-tts-eval-arrow`, EN full set
- Hardware: physical GPU 0, NVIDIA H100 80GB HBM3
- Mode: non-streaming voice cloning, `ref_format=references`
- `c=16 + fp8 + frame_graph + async_decode + compile_sampler`
- Server settings: `max_running_requests=64`, `cuda_graph_max_bs=64`


## Results
| Metric | Candidate |
| --- | ---: |
| Generated requests | 1088/1088 |
| Failed generation requests | 0 |
| Throughput | 5.956 req/s |
| Latency mean / p95 | 2.668 s / 3.864 s |
| RTF mean / p95 | 0.7181 / 0.8248 |
| Corpus WER | 1.31% |
| Corpus WER excluding samples above 50% WER | 1.17% |
| Evaluated / skipped | 1088 / 0 |
| Maximum sample WER | 154.55% |
| Samples above 50% WER | 1 |

**Review by SandyLuXY — APPROVED** [review]
